### DISCUSSION TITLE
How to get charge and discharge status

### DISCUSSION THREAD

**Discussion by vpritulin** [syssi/esphome-jk-bms#609] [Q&A]
Hi, I have charge and discharge enabled on my BMS, but in homeassistant they are marked as disabled.
How can I get their status so that they are shown as enabled?
![image](https://github.com/user-attachments/assets/436fa256-5b8d-4699-b644-16641af92e6c)

**Comment by syssi** [comment]
Could you provide a log of your ESPHome node and the used YAML configuration?

**Reply by vpritulin** [reply]
Log and yaml 
[logs_jk-bms-monitor_logs.txt](https://github.com/user-attachments/files/17329974/logs_jk-bms-monitor_logs.txt)
[bms.txt](https://github.com/user-attachments/files/17329997/bms.txt)

**Reply by vpritulin** [reply]
@syssi maybe I need to add more addresses to get the charge and discharge status?

**Reply by syssi** [reply]
This cannot work. There are no switch and no binary sensor entities exposing the expected details.

**Reply by vpritulin** [reply]
Then it turns out there is no way to get the status?

**Reply by syssi** [reply]
I have to lookup the registers tomorrow. The Modbus example is just a first draft and it's not complete yet.

**Reply by vpritulin** [reply]
@syssi  hey! is there any news on my question?

**Reply by syssi** [reply]
Do you know this document? https://github.com/syssi/esphome-jk-bms/blob/main/docs/pb2a16s20p/BMS%20RS485%20Modbus%20V1.1.pdf

Could you try to retrieve these registers?

![charge-discharge-registers](https://github.com/user-attachments/assets/38ebceea-57ee-4fda-a6e7-3a892e4b4632)

**Reply by vpritulin** [reply]
@syssi thank you! This helped me to get the status of the discharge and discharge. 
But I wanted to know about the status of balancing - is "BalanSta" the status of balancing or is it something else?
<img width="925" alt="Monosnap 极空BMS RS485 Modbus通用协议V1 1 xlsx 2024-10-11 17-10-33" src="https://github.com/user-attachments/assets/8c15f729-4ac1-44be-b1eb-8eb54a0bfd75">

**Reply by syssi** [reply]
Could you give it a try and contribute your new findings?

**Reply by vpritulin** [reply]
I added in the yaml file 

```
  - platform: modbus_controller
    modbus_controller_id: bms0
    name: "status charge"
    address: 0x00C0
    register_type: holding
    value_type: U_WORD
    state_class: measurement

  - platform: modbus_controller
    modbus_controller_id: bms0
    name: "status discharge"
    address: 0x00C1
    register_type: holding
    value_type: U_WORD
    state_class: measurement
```

and modified UI  card

```
- type: markdown
            content: >-
              <center>Discharge:&nbsp;<b>{% if states('sensor.jk_bms_monitor_status_discharge') != '0' %} 
              <font color=#41CD52>{{ 'ON' | upper }}</font> 
              {% else %} 
              <font color=#ff0000>{{ 'OFF' | upper }}</font> 
              {% endif %}</b></center>
          
        - type: markdown
            content: >-
              <center>Discharge:&nbsp;<b>{% if states('sensor.jk_bms_monitor_status_discharge') != '0' %} 
              <font color=#41CD52>{{ 'ON' | upper }}</font> 
              {% else %} 
              <font color=#ff0000>{{ 'OFF' | upper }}</font> 
              {% endif %}</b></center>
```

and got the desired result                     
![image](https://github.com/user-attachments/assets/76a3dfb0-3fea-4560-bedd-fe8b752a82d7)

**Reply by syssi** [reply]
IMO your configuration isn't accurate. The value type `U_WORD` retrieves two bytes starting at register 0x00C0. You have two read the same register (`0x00C0`) in both sensors and use a bitmask to select the first or the second byte.

**Reply by vpritulin** [reply]
@syssi 
do you mean that we need to do something like this?
```
lambda: |-
      // Extract the higher byte (for discharge status)
      return (int(x) >> 8) & 0xFF;

lambda: |-
      // Extract the lower byte (for charge status)
      return int(x) & 0xFF;
```
